pkg/kubelet/prober/worker.go

```diff
@@ -222,6 +222,14 @@ func (w *worker) doProbe() (keepGoing bool) {
 			w.pod.Spec.RestartPolicy != v1.RestartPolicyNever
 	}
 
+	// Graceful shutdown of the pod.
+	if w.pod.ObjectMeta.DeletionTimestamp != nil && w.probeType == liveness {
```

Comment:
should the same logic apply for startup probe? If termination started on a pod that has not started yet, we better stop making startup probes

Reply:
I'm not sure we want to add a special condition for liveness probes + terminating pods. If readiness probes still apply for terminating pods (and they should), I would expect the same to be true for all other probe types.

Reply:
> should the same logic apply for startup probe?

I thought too, but now I don't quite see this happening quite often... having a pod that needs cleaning up and graceful shutdown before having started? If you prefer I can put it back.

Reply:
> I think the issue with liveness probes is the confusion - see the referenced bug description. Pod deletion was requested and it was marked as unhealthy because the code that suppose to check liveness has stopped. And it's not easy to create a liveness probe that will not fail (at least sometimes) on pod termination. Since pod is already being stopped, liveness detection may only change the status of a pod, but will not change anything else in termination logic.
> 
> I think readiness is different as it controls the Load Balancer for the pod. If it will start failing because the code was unloaded, nothing confusing will happen from the customer perspective.

100% agree

Reply:
I see startup probes as a special case of liveness probe, and think they should generally be coupled. This is less likely to happen for a startup probe (how often do you delete a pod before it even starts properly?) but still could, because they use the same code path for container termination. So #52817 and #64715 affect both.